Phylogenetic: fixed clock rate

j23414 opened the pull request:
## Description of proposed changes

Set a clock rate to fix https://github.com/nextstrain/dengue/issues/82

## Related issue(s)

* https://github.com/nextstrain/dengue/issues/82

## Checklist


- [ ] Checks pass
- [ ] Update changelog

j23414 committed:
Phylogenetic: fixed clock rate

In response to review, added fixups:
* In the config, refine and clock_rate are optional
* Add comment on when clock_rates were estimated
* Add entry in CHANGELOG

huddlej reviewed (commented):
Thanks for the quick fix, @j23414! This would work as written, but I had a couple minor comments below that might future-proof this new feature a little.

joverlee521 commented on phylogenetic/defaults/config_dengue.yaml:
```diff
@@ -22,6 +22,24 @@ filter:
+  clock_rate:
+    all:
+      genome: 0.00051
+      E: 0.00052
```
Also interested to know how to define these values for [similar issue in hmpv](https://github.com/nextstrain/hmpv/issues/9)

j23414 replied:
Sure thing! Added a comment noting that the clock rate can be estimated from successful Auspice builds in https://github.com/nextstrain/dengue/pull/123/commits/503075cd01d905399be639c8668019f069a0e43c

This is similar to how the clock rate was fixed for norovirus phylogenetic trees [here](https://github.com/nextstrain/norovirus/issues/22#issuecomment-3221851561).

For example, based on the [hMPV estimated clock rate view](https://next.nextstrain.org/hmpv/all/genome?%402025-12-16&l=clock), I would set this to `--clock-rate 0.000593` for the `all/genome` combination.

Let me know if there’s an easier way to do this, or if I’m interpreting these values incorrectly.

huddlej replied:
Yeah, that's exactly what I would do. Thanks, @j23414!

j23414 commented on phylogenetic/rules/construct_phylogeny.smk:
```diff
@@ -30,6 +30,23 @@ rule tree:
+
+    else leave blank
+    """
+    clock_rate = config['refine']['clock_rate'].get(wildcards.serotype, {}).get(wildcards.gene, "")
```
Thanks! Fixed in https://github.com/nextstrain/dengue/pull/123/commits/152ccaf53413a5063e450de97593875aeff81568 and updated CHANGELOG in https://github.com/nextstrain/dengue/pull/123/commits/c1ca8420da8b65338a7cd67f8ddf75464814749a

I briefly considered moving all [refine parameters](https://github.com/nextstrain/dengue/blob/bdf224d6c67b624ecc35dbea1c9304daf6dc7b52/phylogenetic/rules/construct_phylogeny.smk#L51-L53) and flags into the config, with `refine` required and `clock_rate` optional, but left that for a future refactor, if it becomes necessary.

joverlee521 commented on phylogenetic/rules/construct_phylogeny.smk:
```diff
@@ -30,6 +30,23 @@ rule tree:
+    else leave blank
+    """
+    clock_rate = config.get('refine', {}).get('clock_rate', {}).get(wildcards.serotype, {}).get(wildcards.gene, "")
+    if clock_rate !="":
```
One way to override/remove the default config param is set the param to `null` in a custom config overlay. This check would not catch those cases. Since `null` converts to `None`, would be better to just check for `None` here

```suggestion
    clock_rate = config.get('refine', {}).get('clock_rate', {}).get(wildcards.serotype, {}).get(wildcards.gene, None)
    if clock_rate is not None:
```

j23414 committed:
Fixup: Allow user to override default clock_rate with null

The default clock rate defined in the config can be overridden/removed from an overlay, for example:

refine:
  clock_rate:
    all:
      genome: null

Which may be useful if the clock_rate changes and needs to be reestimated, or the scope of the samples included changes.

Co-authored-by: Jover Lee <joverlee521@gmail.com>
